dask/utils.py

```diff
@@ -166,75 +186,81 @@ def get_bin_linesep(encoding, linesep):
+              linesep=os.linesep):
+    """Pull out a block of text from a file given start and stop bytes.
 
-    This gets data starting/ending from the next linesep delimiter
+    This gets data starting/ending from the next linesep delimiter.
 
     Examples
     --------
-
-    >> with open('myfile.txt', 'w') as f:
+    >> with open('myfile.txt', 'wb') as f:
     ..     f.write('123\n456\n789\nabc')
 
-    >> f = open('myfile.txt')
-
-    In the example below, 1 and 10 don't line up with endlines
+    In the example below, 1 and 10 don't line up with endlines.
 
-    >> textblock(f, 1, 10)
+    >> u''.join(textblock('myfile.txt', 1, 10))
     '456\n789\n'
     """
-    if isinstance(file, (str, unicode)):
-        myopen = opens.get(compression, open)
-        f = myopen(file, 'rb')
-        try:
-            result = textblock(f, start, stop, compression=None,
-                               encoding=encoding, linesep=linesep)
-        finally:
-            f.close()
-        return result
+    # Make sure `linesep` is not a byte string because `io.TextIOWrapper` in
+    # Python versions other than 2.7 dislike byte strings for the `newline`
+    # argument.
+    linesep = str(linesep)
 
-    if encoding is None:
-        encoding = getattr(file, 'encoding', 'utf-8')
-        if encoding is None:
-            encoding = 'utf-8'
-
-    if linesep is None:
-        linesep = os.linesep
-
-    if start:
-        startstart, startstop = next_linesep(file, start, encoding, linesep)
-    else:
-        startstart = start
-        startstop = start
-
-    if stop is None:
-        file.seek(start)
-        return file.read()
-
-    stopstart, stopstop = next_linesep(file, stop, encoding, linesep)
-
-    file.seek(startstop)
-
-    return file.read(stopstop - startstop)
+    chunksize = end - start + 1
+    f = open(filename, 'rb', compression)
+    try:
+        with io.BufferedReader(f) as fb:
+            # Get byte representation of the line separator.
+            bin_linesep = get_bin_linesep(encoding, linesep)
+            bin_linesep_len = len(bin_linesep)
+
+            # If `start` does not correspond to the beginning of the file, then
+            # we need to set the file pointer to `start` and keep reading
+            # chunks until we find a line separator. We then set the file
+            # pointer to the position after the line separator.
+            # If `start` corresponds to the first character of a line, the line
+            # is not part of this textblock. This is okay since
+            # io.TextIOWrapper below keeps reading lines until `start` exceeds
+            # `end`, i.e also the line where `start == end` is part of this
+            # textblock. Thus, the line is definitely only part of one
+            # textblock.
+            if start > 0:
+                fb.seek(start)
+                while True:
+                    buf = f.read(4096)
```

Comment:
I think there is a corner case where the line separator is ``\r\n`` and ``f.read(4096)`` reads up to ``\r``. Then, the line separator is not detected.